File: crates/wadup-guest/src/table.rs
use crate::ffi;
use crate::types::{Column, DataType, Value};

pub struct Table {
    name: String,
}

impl Table {
    pub fn define(name: impl Into<String>, columns: Vec<(&str, DataType)>) -> Result<Self, String> {
        let name = name.into();

        let cols: Vec<Column> = columns
            .into_iter()
            .map(|(n, t)| Column {
                name: n.to_string(),
                data_type: t,
            })
            .collect();

        let columns_json = serde_json::to_string(&cols)
            .map_err(|e| format!("Failed to serialize columns: {}", e))?;

        unsafe {
            let result = ffi::define_table(
                name.as_ptr(),
                name.len(),
                columns_json.as_ptr(),
                columns_json.len(),
            );

            if result < 0 {
                return Err(format!("Failed to define table '{}'", name));
            }
        }

        Ok(Table { name })
    }

    pub fn insert(&self, values: &[Value]) -> Result<(), String> {
        let values_json = serde_json::to_string(values)
            .map_err(|e| format!("Failed to serialize values: {}", e))?;

        unsafe {
            let result = ffi::insert_row(
                self.name.as_ptr(),
                self.name.len(),
                values_json.as_ptr(),
                values_json.len(),
            );

            if result < 0 {
                return Err(format!("Failed to insert row into '{}'", self.name));
            }
        }

        Ok(())
    }
}

pub struct TableBuilder {
    name: String,
    columns: Vec<(&'static str, DataType)>,
}

impl TableBuilder {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            columns: Vec::new(),
        }
    }

    pub fn column(mut self, name: &'static str, data_type: DataType) -> Self {
        self.columns.push((name, data_type));
        self
    }

    pub fn build(self) -> Result<Table, String> {
        Table::define(self.name, self.columns)
    }
}

File: go-wadup-guest/table.go
package wadup

import (
	"encoding/json"
	"fmt"
)

// Table represents a defined table that can accept row insertions
type Table struct {
	name string
}

// DefineTable defines a new table with the given columns
func DefineTable(name string, columns []Column) (*Table, error) {
	// Serialize columns to JSON
	columnsJSON, err := json.Marshal(columns)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize columns: %w", err)
	}

	// Call FFI
	namePtr, nameLen := stringToFFI(name)
	colPtr, colLen := stringToFFI(string(columnsJSON))

	result := defineTableFFI(namePtr, nameLen, colPtr, colLen)
	if result < 0 {
		return nil, fmt.Errorf("failed to define table '%s': error code %d", name, result)
	}

	return &Table{name: name}, nil
}

// InsertRow inserts a row of values into the table
func (t *Table) InsertRow(values []Value) error {
	// Serialize values to JSON
	valuesJSON, err := json.Marshal(values)
	if err != nil {
		return fmt.Errorf("failed to serialize values: %w", err)
	}

	// Call FFI
	namePtr, nameLen := stringToFFI(t.name)
	valPtr, valLen := stringToFFI(string(valuesJSON))

	result := insertRowFFI(namePtr, nameLen, valPtr, valLen)
	if result < 0 {
		return fmt.Errorf("failed to insert row into '%s': error code %d", t.name, result)
	}

	return nil
}

// TableBuilder provides a fluent API for building tables
type TableBuilder struct {
	name    string
	columns []Column
}

// NewTableBuilder creates a new table builder
func NewTableBuilder(name string) *TableBuilder {
	return &TableBuilder{
		name:    name,
		columns: make([]Column, 0),
	}
}

// Column adds a column to the table
func (b *TableBuilder) Column(name string, dataType DataType) *TableBuilder {
	b.columns = append(b.columns, Column{
		Name:     name,
		DataType: dataType,
	})
	return b
}

// Build creates the table
func (b *TableBuilder) Build() (*Table, error) {
	return DefineTable(b.name, b.columns)
}
